cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/SnapshotSink.java

```diff
@@ -0,0 +1,118 @@
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.collect.Maps;
+
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ *  CDAP Snapshot Dataset Batch Sink.
+ */
+@Plugin(type = "sink")
+@Name("SnapshotSink")
+@Description("CDAP Snapshot Dataset Batch Sink")
+public class SnapshotSink extends BatchWritableSink<StructuredRecord, byte[], Put> {
+
+  private static final String NAME_DESC = "Name of the snapshot dataset. If it does not already exist, one will be " +
+    "created.";
+  private static final String PROPERTY_SCHEMA_DESC = "Optional schema of the snapshot dataset as a JSON Object. " +
+    "If the dataset does not already exist, one will be created with this schema, which will allow the table to be " +
+    "explored through Hive.\"";
+  private static final String PROPERTY_SCHEMA_ROW_FIELD_DESC = "The name of the record field that should be used as " +
+    "the row key when writing to the snapshot dataset.";
+
+  private RecordPutTransformer recordPutTransformer;
+  private final SnapshotSinkConfig snapshotSinkConfig;
+
+  public SnapshotSink(SnapshotSinkConfig snapshotSinkConfig) {
+    this.snapshotSinkConfig = snapshotSinkConfig;
+  }
+
+  @Override
+  public void initialize(BatchSinkContext context) throws Exception {
+    super.initialize(context);
+    Map<String, String> sinkArgs = new HashMap<>();
+    SnapshotDataset snapshotDataset = context.getDataset(snapshotSinkConfig.name, sinkArgs);
+    snapshotDataset.updateMetaDataTable(snapshotDataset.getTransactionId());
+    recordPutTransformer = new RecordPutTransformer(snapshotSinkConfig.rowField,
+      Schema.parseJson(snapshotSinkConfig.schemaStr), true);
+  }
+
+  @Override
+  protected Map<String, String> getProperties() {
+    Map<String, String> properties = Maps.newHashMap(snapshotSinkConfig.getProperties().getProperties());
+    properties.put(Properties.BatchReadableWritable.NAME, snapshotSinkConfig.name);
+    properties.put(Properties.BatchReadableWritable.TYPE, SnapshotDataset.class.getName());
+    return properties;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(snapshotSinkConfig.rowField),
+      "Row field must be given as a property.");
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<KeyValue<byte[], Put>> emitter) throws Exception {
+    Put put = recordPutTransformer.toPut(input);
+    emitter.emit(new KeyValue<>(put.getRow(), put));
+  }
+
+  /**
+   *  Config for the snapshot sink
+   */
+  public static class SnapshotSinkConfig extends PluginConfig {
+    @Description(NAME_DESC)
+    private String name;
+
+    @Name(Properties.Table.PROPERTY_SCHEMA)
+    @Description(PROPERTY_SCHEMA_DESC)
+    @Nullable
+    String schemaStr;
```

Comment:
if it is nullable, it should have a default value. Assign in no-args constructor.